### PR TITLE
[DAPS-1331] Collection ID Updated on restart

### PR DESCRIPTION
**Pull request by AronPerez**
## Ticket  

[DAPS-1331](https://github.com/ORNL/DataFed/issues/1331)

## Description 

The original code in `scripts/globus/create_guest_collection.py` (lines 166-168) always deleted and recreated guest collections on container restart, even when the base path hadn't changed, causing unnecessary service disruptions. To address this issue, I modified the script to delete and recreate guest collections only when the base path actually changes. 

## Technical
                                                                                                                                                                                                                                    
1. **Enhanced Collection Discovery**: Added logic to capture the existing base path when finding guest collections                                                                                                                  
2. **Smart Deletion Logic**: Added conditional logic that only deletes collections when:                                                                                                                                            
   - The base path has actually changed, OR                                                                                                                                                                                         
   - The existing base path cannot be determined (safe fallback)                                                                                                                                                                    
3. **Conditional Collection Creation**: Only creates new collections when needed (when no existing collection is found)    

## How Has This Been Tested?  


<details><summary>Using Compose All</summary>
<p>

1. `cd compose/all`
2. Follow the setup for compose all
3. `
docker compose up -d`
4. `docker logs datafed-gcs`
5. Take note of the `End-point ID:
6. `docker compose down`
7. `docker compose up -d`
8. `docker logs datafed-gcs`

</p>
</details> 



## Artifacts (if appropriate):  

### Globus Cred Path File

<details><summary>DATAFED_GLOBUS_CRED_FILE_PATH for Cache dir</summary>
<p>

```bash
DATAFED_GLOBUS_CRED_FILE_PATH:      /opt/datafed/globus/client_cred.json
```

</p>
</details> 



<details><summary>Restarting Service</summary>
<p>



```bash
Smokey_Mountain Collection Guest
Found existing guest collection d2e777c6-7413-4ce0-84f0-89de89182cdf
Collection data keys: ['DATA_TYPE', 'authentication_timeout_mins', 'collection_type', 'connector_id', 'contact_email', 'created_at', 'disable_anonymous_writes', 'disable_verify', 'display_name', 'domain_name', 'enable_https', 'force_encryption', 'force_verify', 'high_assurance', 'https_url', 'id', 'identity_id', 'last_access', 'manager_url', 'mapped_collection_id', 'policies', 'public', 'require_mfa', 'restrict_transfers_to_high_assurance', 'storage_gateway_id', 'subscription_admin_verified', 'tlsftp_url', 'user_credential_id']
Found existing guest collection d2e777c6-7413-4ce0-84f0-89de89182cdf
Found cached base path: '/'
Current BASE_PATH environment variable: '/'
Decision: Base path unchanged ('/'), keeping existing guest collection
b80380c3-84a3-41a7-96ba-1c8067550d5f
Using existing guest collection d2e777c6-7413-4ce0-84f0-89de89182cdf
Cached base path '/' for collection d2e777c6-7413-4ce0-84f0-89de89182cdf
Successfully cached base path for future comparisons
```

</p>
</details>

### CODE REVIEW

**Comment by JoshuaSBrown** [comment]
I'm guessing you already know this, just putting here for tracking, currently the CI is still recreating the guest collection. 

image

So it looks like something with the base path comparison is not quite right.

**Comment by AronPerez** [comment]
The Globus API doesn't return base path information for guest collections, making it impossible to determine if the base path has changed. This was causing unnecessary recreation of guest collections on every restart.                                                                                                                  

Solution:
- Use a local cache file to store the base path used when creating collections
- Compare cached base path with current BASE_PATH environment variable
- Only recreate collection when base path actually changes
- Add DATAFED_FORCE_RECREATE_GUEST_COLLECTION override for manual recreation
- Clean up cache files when collections are deleted

File: scripts/globus/create_guest_collection.py [hunk 1]
```diff
@@ -155,17 +155,48 @@
...
         break
 
+# Only delete and recreate the guest collection if the base path has changed
+should_recreate_collection = False
 if guest_collection_found:
+    if existing_base_path is None:
+        # If we can't get the base path from the list, try to get collection details
+        try:
+            collection_details = client.get_collection(guest_collection_id)
+            existing_base_path = collection_details.get("collection_base_path")
```
[the remark this answers is not in the document]
Comment (AronPerez) [reply scripts/globus/create_guest_collection.py]: Guest collections in Globus have different metadata exposure than mapped collections. The `collection_base_path is:
  - Set during creation via `GuestCollectionDocument(collection_base_path=BASE_PATH)`
  - Used internally by Globus for path resolution
  - Not returned in the public API responses for guest collections

  This is likely a Globus design decision because:
  - Guest collections are meant to be more restricted/controlled 🚧 
  - The base path is considered internal configuration
  - Exposing it might create security or operational concerns 😨

**Comment by AronPerez** [comment]
## Summary by Sourcery

Improve guest collection management in the Globus script by only deleting and recreating collections when the base path actually changes or cannot be determined, avoiding unnecessary service disruptions.

Bug Fixes:
- Prevent unnecessary deletion and recreation of guest collections on container restart when the base path is unchanged

Enhancements:
- Capture existing collection base path from list or API details
- Add conditional logic to determine when to delete and recreate collections based on base path comparison
- Only create a new guest collection if none exists or after deletion due to base path change

File: scripts/globus/create_guest_collection.py [hunk 2]
```diff
@@ -151,23 +151,123 @@
...
+    cache_dir = os.path.dirname(CRED_FILE_PATH)
+    should_recreate_collection, recreate_reason = should_recreate_guest_collection(
+        guest_collection_id, BASE_PATH, cache_dir
+    )
+    
+    print(f"Decision: {recreate_reason}")
+    if should_recreate_collection:
+        print("Action: Will recreate guest collection")
+
+def recreate_guest_collection(client, collection_id, cache_directory):
```
Comment (AronPerez) [inline scripts/globus/create_guest_collection.py:230]: To prevent name shadowing, I've simplified the parameters and am allowing the comments and functions to clarify

File: scripts/globus/create_guest_collection.py [hunk 3]
```diff
@@ -151,23 +151,123 @@
...
+        # We'll need to get it from collection details
+        existing_base_path = None
+        print(f"Found existing guest collection {guest_collection_id}")
+        print(f"Collection data keys: {list(item.keys())}")
         break
 
-if guest_collection_found:
-    print(f"Removing current guest collection {guest_collection_id}")
-    response = client.delete_collection(guest_collection_id)
+def should_recreate_guest_collection(collection_id, current_base_path, cache_directory):
```
Comment (AronPerez) [inline scripts/globus/create_guest_collection.py:173]: https://github.com/ORNL/DataFed/pull/1468/files#r2131429298

File: scripts/globus/create_guest_collection.py [hunk 4]
```diff
@@ -183,20 +283,66 @@
         identity_id=client_id,
         username=local_username,
     )
-    client.create_user_credential(credential_document)
-
-# We are recreating the collection from scratch because we are unable to update
-# the base path once the collection has been made this is a limitation of Globus
-# Create the collection
-collection_document = globus_sdk.GuestCollectionDocument(
-    public="True",
```
Comment (AronPerez) [inline scripts/globus/create_guest_collection.py:192]: `public` is supposed to be a `bool`

**Comment by AronPerez** [comment]
Test Scenarios

  1. Base Path Change Detection

  Scenario 1.1: First-time Collection Creation
  - Precondition: No existing guest collection or cache file
  - Test: Set `DATAFED_GCS_COLLECTION_BASE_PATH`="/data" and run script
  - Expected: Creates new guest collection with base path "/data" and caches the
  path

  Scenario 1.2: Base Path Unchanged on Restart
  - Precondition: Existing guest collection with cached base path "/data"
  - Test: Restart with same `DATAFED_GCS_COLLECTION_BASE_PATH`="/data"
  - Expected: Keeps existing collection, logs "Base path unchanged, keeping
  existing guest collection"

  Scenario 1.3: Base Path Changed
  - Precondition: Existing guest collection with cached base path "/data"
  - Test: Change `DATAFED_GCS_COLLECTION_BASE_PATH` to "/new-data"
  - Expected: Deletes old collection, creates new one with "/new-data", updates
  cache

  2. Cache File Management

  Scenario 2.1: Missing Cache File
  - Precondition: Existing guest collection but no cache file
  - Test: Run script without cache file
  - Expected: Keeps existing collection to avoid unnecessary recreation, logs
  appropriate message

  Scenario 2.2: Corrupted Cache File
  - Precondition: Cache file exists but contains invalid data
  - Test: Corrupt cache file and run script
  - Expected: Handles gracefully, treats as missing cache, keeps existing
  collection

  Scenario 2.3: Cache File Permissions
  - Precondition: Cache directory has restricted permissions
  - Test: Run script with read-only cache directory
  - Expected: Logs warning about cache failure but continues operation

  3. Path Normalization

  Scenario 3.1: Trailing Slash Handling
  - Test Cases:
    - Cache: "/data", Current: "/data/" → Should not recreate
    - Cache: "/data/", Current: "/data" → Should not recreate
    - Cache: "/", Current: "/" → Should not recreate

  Scenario 3.2: Empty vs Root Path
  - Test: Switch between empty string and "/" as base path
  - Expected: Proper normalization and appropriate recreation logic

  4. Force Recreation Override

  Scenario 4.1: Manual Override
  - Precondition: Existing collection with unchanged base path
  - Test: Set DATAFED_FORCE_RECREATE_GUEST_COLLECTION=true
  - Expected: Forces recreation regardless of base path comparison

  Scenario 4.2: Override with Path Change
  - Precondition: Existing collection with changed base path
  - Test: Set force override flag
  - Expected: Recreates collection (would happen anyway due to path change)

  5. Error Handling

  Scenario 5.1: Collection Deletion Failure
  - Test: Simulate Globus API failure during collection deletion
  - Expected: Logs error, continues with existing collection, doesn't update cache

  Scenario 5.2: Collection Creation Failure
  - Test: Simulate failure during new collection creation
  - Expected: Exits with error code 1, doesn't create invalid cache

  Scenario 5.3: Globus API Timeout
  - Test: Simulate network timeout during API calls
  - Expected: Appropriate error handling and recovery

  6. Cache Consistency

  Scenario 6.1: Cache Update After Creation
  - Test: Create new collection and verify cache file contains correct base path
  - Expected: Cache file exists with exact base path used for collection

  Scenario 6.2: Cache Cleanup After Deletion
  - Test: Delete collection and verify old cache file is removed
  - Expected: Old cache file is properly cleaned up

  7. Edge Cases

  Scenario 7.1: Multiple Collections
  - Test: Run with multiple guest collections having same display name
  - Expected: Handles collection identification correctly

  Scenario 7.2: Concurrent Execution
  - Test: Run multiple instances of script simultaneously
  - Expected: Proper file locking or race condition handling

  Scenario 7.3: Very Long Base Paths
  - Test: Use extremely long base path values
  - Expected: Handles without truncation or errors

  8. Integration Testing

  Scenario 8.1: Full DataFed Restart Cycle
  - Test: Complete DataFed system restart with existing collections
  - Expected: No unnecessary collection recreation, all services work correctly

  Scenario 8.2: Configuration Migration
  - Test: Migrate from old DataFed version without caching to new version
  - Expected: Smooth transition, appropriate handling of missing cache

  9. Performance Testing

  Scenario 9.1: Large Number of Collections
  - Test: Environment with many existing collections
  - Expected: Efficient collection identification and processing

  Scenario 9.2: Frequent Restarts
  - Test: Multiple rapid restarts with unchanged configuration
  - Expected: Consistent behavior, no collection thrashing

  10. Validation Testing

  Scenario 10.1: Base Path Format Validation
  - Test: Invalid base path formats (e.g., relative paths, invalid characters)
  - Expected: Proper validation and error messages

  Scenario 10.2: Environment Variable Validation
  - Test: Missing or invalid required environment variables
  - Expected: Clear error messages and graceful failure

  Test Data Requirements

  - Various base path configurations
  - Mock Globus API responses
  - Different cache file states
  - Multiple environment variable combinations

  Success Criteria

  - No unnecessary guest collection recreation on restart
  - Proper base path change detection
  - Robust error handling and recovery
  - Consistent cache management
  - Backwards compatibility with existing deployments

File: scripts/globus/create_guest_collection.py [hunk 5]
```diff
@@ -151,23 +151,123 @@
...
+    print(f"Current BASE_PATH environment variable: '{current_base_path}'")
+    
+    # Check for manual override first
+    force_recreate = os.getenv("DATAFED_FORCE_RECREATE_GUEST_COLLECTION", "false").lower() == "true"
+    if force_recreate:
+        return True, "DATAFED_FORCE_RECREATE_GUEST_COLLECTION is set, forcing recreation"
+    
+    # Compare base paths
+    if cached_base_path is None:
+        return False, "No cached base path available, keeping existing collection to avoid unnecessary recreation"
```
Comment (JoshuaSBrown) [inline scripts/globus/create_guest_collection.py:206]: Wouldn't you want to recreate by default here? If the base path file is not available you don't know if the base path is the same, the only way to guarantee that is to recreate it.

**Review by JoshuaSBrown — APPROVED** [review]
Other than the one comment I left this will solve the bug. Ideally, we will have all the configuration in a single file for instance the .json file where the client_creds are stored would be fine, using a .txt file for the base path is adequate. The CI is already passing for with out recreation I would like you to run it with changing the ENV variable and get a print screen of it forcibly recreating the guest collection as well.

image
